"Toggling" IT: dev-test

Hi!

Not sure it is related to maybe #1626 (PR not merged yet), but this test will sometimes fail:

`DevTest.configIncludesChangeTest()`:

```
[INFO] [INFO] --- maven-jar-plugin:3.2.2:jar (default-jar) @ dev-it ---
Warning: ARNING] JAR will be empty - no content was marked for inclusion!
[INFO] [INFO] Building jar: /home/runner/work/ci.maven/ci.maven/liberty-maven-plugin/target/it/dev-it/target/dev-it-1.0-SNAPSHOT.jar
[INFO] [INFO] 
[INFO] [INFO] --- maven-jar-plugin:3.2.2:test-jar (default) @ dev-it ---
[INFO] [INFO] Building jar: /home/runner/work/ci.maven/ci.maven/liberty-maven-plugin/target/it/dev-it/target/dev-it-1.0-SNAPSHOT-tests.jar
[INFO] [INFO] 
[INFO] [INFO] --- maven-failsafe-plugin:3.0.0-M7:integration-test (integration-test) @ dev-it ---
[INFO] [INFO] Using auto detected provider org.apache.maven.surefire.junit4.JUnit4Provider
[INFO] [INFO] 
[INFO] [INFO] -------------------------------------------------------
[INFO] [INFO]  T E S T S
[INFO] [INFO] -------------------------------------------------------
[INFO] [INFO] Running net.wasdev.wlp.test.dev.it.MultiModuleRunInstallEarTest
[INFO] [INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 52.165 s - in net.wasdev.wlp.test.dev.it.MultiModuleRunInstallEarTest
[INFO] [INFO] Running net.wasdev.wlp.test.dev.it.DevCopyTestDependenciesTest
[INFO] [INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 26.323 s - in net.wasdev.wlp.test.dev.it.DevCopyTestDependenciesTest
[INFO] [INFO] Running net.wasdev.wlp.test.dev.it.MultipleConcurrentLibertyModulesPlTest
[INFO] [INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 40.805 s - in net.wasdev.wlp.test.dev.it.MultipleConcurrentLibertyModulesPlTest
[INFO] [INFO] Running net.wasdev.wlp.test.dev.it.MultiModuleRunEjbTest
[INFO] [INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 40.328 s - in net.wasdev.wlp.test.dev.it.MultiModuleRunEjbTest
[INFO] [INFO] Running net.wasdev.wlp.test.dev.it.MultiModuleHotTestingTest
[INFO] [INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 61.915 s - in net.wasdev.wlp.test.dev.it.MultiModuleHotTestingTest
[INFO] [INFO] Running net.wasdev.wlp.test.dev.it.DevTest
Error: [ERROR] Tests run: 13, Failures: 1, Errors: 0, Skipped: 1, Time elapsed: 127.036 s <<< FAILURE! - in net.wasdev.wlp.test.dev.it.DevTest
Error: [ERROR] net.wasdev.wlp.test.dev.it.DevTest.configIncludesChangeTest  Time elapsed: 23.799 s  <<< FAILURE!

…

[INFO] java.lang.AssertionError
[INFO] 	at org.junit.Assert.fail(Assert.java:87)
[INFO] 	at org.junit.Assert.assertTrue(Assert.java:42)
[INFO] 	at org.junit.Assert.assertTrue(Assert.java:53)
[INFO] 	at net.wasdev.wlp.test.dev.it.DevTest.configIncludesChangeTest(DevTest.java:107)
[INFO] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[INFO] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[INFO] 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[INFO] 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
[INFO] 	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
```

Line 107 is:

`replaceString("<feature>servlet-4.0</feature>", "<!-- replace -->", srcServerXMLIncludes);`


Assuming we are looking at the wrong line now.
Looking at the test, it depends on timeouts:

`boolean foundUpdate = verifyLogMessageExists("<feature>servlet-4.0</feature>", 60000, targetServerXMLIncludes);`

… which is 60 seconds. Maybe you might want to use [Awaitility](http://www.awaitility.org) here? It would give you a much nicer exception message "for free"..

It would read to:


```java
      await()
        .atMost(Duration.ofSeconds(60L))
        .until(() -> verifyLogMessageExists("<feature>servlet-4.0</feature>", 10, targetServerXMLIncludes));
```

(10 ms chosen randomly).

This would help to identify the actual problem.

@scottkurz @cherylking  Some other tests seem to fail, too. For example, in the [current main build](https://pipelines.actions.githubusercontent.com/serviceHosts/00639a69-f5fb-4468-907f-d57b48d71543/_apis/pipelines/1/runs/1183/signedlogcontent/16?urlExpires=2023-02-13T18%3A54%3A39.4710637Z&urlSigningMethod=HMACV1&urlSignature=q6h2DzGm395p03p7gn0jYdgkA%2FR6O56%2FS5yYbYmJKHw%3D), there's this:

```
2023-02-01T20:33:52.2556285Z [INFO] [ERROR] Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 59.349 s <<< FAILURE! - in net.wasdev.wlp.test.dev.it.MultiModuleHotTestingTest
2023-02-01T20:33:52.2569955Z [INFO] [ERROR] net.wasdev.wlp.test.dev.it.MultiModuleHotTestingTest.runTest  Time elapsed: 37.299 s  <<< FAILURE!
2023-02-01T20:33:52.2570665Z [INFO] java.lang.AssertionError: 
2023-02-01T20:33:52.2571546Z [INFO] Last 100 lines of log at /home/runner/work/ci.maven/ci.maven/liberty-maven-plugin/target/it/dev-it/target/../resources/multi-module-projects/typeA/logFile.txt:
2023-02-01T20:33:52.2572098Z [INFO] ===================== START =======================
2023-02-01T20:33:52.2572695Z [INFO] 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:122)
2023-02-01T20:33:52.2573400Z [INFO] 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:80)
2023-02-01T20:33:52.2573957Z [INFO] 	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)

[...]

2023-02-01T20:33:52.2599835Z [INFO] [INFO] 
2023-02-01T20:33:52.2600114Z [INFO] [INFO] Results:
2023-02-01T20:33:52.2600394Z [INFO] [INFO] 
2023-02-01T20:33:52.2600675Z [INFO] [ERROR] Failures: 
2023-02-01T20:33:52.2601061Z [INFO] [ERROR]   ConverterUnitTest.testHeightFeet:12 expected: <2> but was: <4>
2023-02-01T20:33:52.2601442Z [INFO] [INFO] 
2023-02-01T20:33:52.2601780Z [INFO] [ERROR] Tests run: 1, Failures: 1, Errors: 0, Skipped: 0
2023-02-01T20:33:52.2602102Z [INFO] [INFO] 
2023-02-01T20:33:52.2602434Z [INFO] [ERROR] Unit tests failed: There are test failures.
```

Sadly, it seems to be hard to execute the profile `online-its` locally, see:

https://github.com/OpenLiberty/ci.maven/blob/main/.github/workflows/maven.yml#L59-L69

So, what's the next best way to proceed here?

So, looking again at the tests, I see:

```
2023-05-04T01:28:03.4175956Z [INFO] [ERROR] Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 25.536 s <<< FAILURE! - in net.wasdev.wlp.test.dev.it.MultiModuleRunInstallEarTest
2023-05-04T01:28:03.4186009Z [INFO] [ERROR] net.wasdev.wlp.test.dev.it.MultiModuleRunInstallEarTest.notOverwriteExistingM2Test  Time elapsed: 0.015 s  <<< FAILURE!
```

Now there's a catch here. I found `forkMode=once` in your test setup which is not supported anymore.
Instead use 
```
                    <forkCount>1</forkCount>
                    <reuseForks>true</reuseForks>
```

Same for  `generate-features-it`.
I wonder it only fails on OpenLiberty. Maybe the runtime does not have the requested features anymore?
Anyway, both fail on `newFeatureFile` not exists.

--- a/liberty-maven-plugin/src/it/dev-it/pom.xml
+++ b/liberty-maven-plugin/src/it/dev-it/pom.xml
@@ -49,10 +49,11 @@
             <plugin>
                 <groupId>org.apache.maven.plugins</groupId>
                 <artifactId>maven-failsafe-plugin</artifactId>
-                <version>3.0.0-M7</version>
+                <version>3.0.0</version>
                 <configuration>
                     <redirectTestOutputToFile>true</redirectTestOutputToFile>
-                    <forkMode>once</forkMode>
+                    <forkCount>1</forkCount>
+                    <reuseForks>true</reuseForks>
                     <forkedProcessTimeoutInSeconds>2400</forkedProcessTimeoutInSeconds>
                     <argLine>-enableassertions</argLine>
                     <workingDirectory>${project.build.directory}</workingDirectory>
@@ -84,7 +85,7 @@
             <plugin>
                 <groupId>org.apache.maven.plugins</groupId>
                 <artifactId>maven-surefire-plugin</artifactId>
-                <version>3.0.0-M7</version>
+                <version>3.0.0</version>
                 <configuration>
                     <skip>true</skip>
                 </configuration>

--- a/liberty-maven-plugin/src/it/generate-features-it/pom.xml
+++ b/liberty-maven-plugin/src/it/generate-features-it/pom.xml
@@ -49,10 +49,11 @@
             <plugin>
                 <groupId>org.apache.maven.plugins</groupId>
                 <artifactId>maven-failsafe-plugin</artifactId>
-                <version>2.19.1</version>
+                <version>3.0.0</version>
                 <configuration>
                     <redirectTestOutputToFile>true</redirectTestOutputToFile>
-                    <forkMode>once</forkMode>
+                    <forkCount>1</forkCount>
+                    <reuseForks>true</reuseForks>
                     <forkedProcessTimeoutInSeconds>1200</forkedProcessTimeoutInSeconds>
                     <argLine>-enableassertions</argLine>
                     <workingDirectory>${project.build.directory}</workingDirectory>
@@ -84,7 +85,7 @@
             <plugin>
                 <groupId>org.apache.maven.plugins</groupId>
                 <artifactId>maven-surefire-plugin</artifactId>
-                <version>2.19.1</version>
+                <version>3.0.0</version>
                 <configuration>
                     <skip>true</skip>
                 </configuration>

